### PR TITLE
Fixes overflow trap when creating DispatchTime objects with large upt…

### PR DESCRIPTION
**Pull request by ktopley-apple**
…imeNanoseconds values and re-enables the tests that failed.

Adds some additional tests.

(rdar://problem/34751238)

The change is in three parts:

1. The value UInt64.max passed to the DispatchTime initializer means DISPATCH_TIME_FOREVER and should not be scaled using the Mach timebase.
2. The scaling calculations need to detect overflow and clamp to UInt64.max if this occurs, making the value equal to DISPATCH_TIME_FOREVER
3. The getter for the uptimeNanoseconds also needs to be protected against overflow and return UInt64.max if it occurs.

### CODE REVIEW

**Comment by ktopley-apple** [comment]
@swift-ci please clean smoke test

File: stdlib/public/SDK/Dispatch/Time.swift [hunk 1]
```diff
@@ -53,17 +53,26 @@ public struct DispatchTime : Comparable {
 	///               system sleep time), not zero nanoseconds since boot.
 	public init(uptimeNanoseconds: UInt64) {
 		var rawValue = uptimeNanoseconds
-		if (DispatchTime.timebaseInfo.numer != DispatchTime.timebaseInfo.denom) {
-			rawValue = (rawValue * UInt64(DispatchTime.timebaseInfo.denom) 
-				+ UInt64(DispatchTime.timebaseInfo.numer - 1)) / UInt64(DispatchTime.timebaseInfo.numer)
+
+		// UInt64.max means distantFuture. Do not try to scale it.
+		if rawValue != UInt64.max && DispatchTime.timebaseInfo.numer != DispatchTime.timebaseInfo.denom {
```
Comment (jrose-apple) [inline stdlib/public/SDK/Dispatch/Time.swift]: It seems like the scaling could always trigger an overflow, right? Is this sufficient?

Comment (ktopley-apple) [reply stdlib/public/SDK/Dispatch/Time.swift]: I protected the operations with overflow checks, apart from the division which can't overflow, so I don't think anything else can trap.

Comment (gparker42) [reply stdlib/public/SDK/Dispatch/Time.swift]: So what is the `if` for? Is it simply trying to optimize some values on some platforms?

Comment (ktopley-apple) [reply stdlib/public/SDK/Dispatch/Time.swift]: The first part lets UInt64.max get through unchanged, the second part avoids the arithmetic when there is no timebase scaling. So yes, that part is an optimization (that was already there).

File: test/stdlib/Dispatch.swift [hunk 2]
```diff
@@ -119,6 +115,17 @@ DispatchAPI.test("DispatchTime comparisons") {
...
+	var time = DispatchTime(uptimeNanoseconds: 0)
+	expectEqual(time.uptimeNanoseconds, 0)
+
+	time = DispatchTime(uptimeNanoseconds: 15 * NSEC_PER_SEC)
+	expectEqual(time.uptimeNanoseconds, 15 * NSEC_PER_SEC)
+
+	time = DispatchTime(uptimeNanoseconds: UInt64.max)
+	expectEqual(time.uptimeNanoseconds, UInt64.max)
+}
+
```
Comment (gparker42) [inline test/stdlib/Dispatch.swift]: I don't think any of these exercise the overflow paths. You should add values like `UInt64.max-1`.

Comment (ktopley-apple) [reply test/stdlib/Dispatch.swift]: That's a good point and when I tried some of the larger values, it turns out that the forced ceiling of UInt64.max for the multiplication produces some odd results. I am going to revisit that.

**Comment by ktopley-apple** [comment]
@swift-ci please clean smoke test

**Comment by ktopley-apple** [comment]
@swift-ci please clean smoke test

**Review by mwwa — APPROVED** [review]

**Comment by ktopley-apple** [comment]
Squashed commits.

**Comment by ktopley-apple** [comment]
@swift-ci please smoke test

**Pull request merged**